### PR TITLE
Initial public draft for TotW 99

### PR DESCRIPTION
**Pull request by tituswinters**

### CODE REVIEW

File: _posts/2017-11-16-totw-99.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+Revised 2017-10-10
+
+The interface of a C++ class is not constrained to its members or to its
+definition. When evaluating an API, we must consider definitions beyond the
+class body that can be as much a part of its interface as its public members.
+
+These external interface points include template specializations like hashers or
+traits, nonmember operator overloads (e.g. logging, relationals), and other
+canonical nonmember functions designed for use with argument-dependent lookup
+(ADL), most notably `swap`.
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: swap()

File: _posts/2017-11-16-totw-99.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+struct hash<space::Key> {
+  size_t operator()(const space::Key& x) const;
+};
+HASH_NAMESPACE_END
+```
+
+There are some important risks associated with making such extensions
+incorrectly, so this article will try to present some guidance.
+
+### The proper namespace
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: Initial Caps Headings

File: _posts/2017-11-16-totw-99.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+class body that can be as much a part of its interface as its public members.
+
+These external interface points include template specializations like hashers or
+traits, nonmember operator overloads (e.g. logging, relationals), and other
+canonical nonmember functions designed for use with argument-dependent lookup
+(ADL), most notably `swap`.
+
+A few of these are illustrated below for some sample class `space::Key`:
+
+```c++
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: I think we should standardize on indent 2 spaces for tips. Any issues with that?

File: _posts/2017-11-16-totw-99.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+```
+
+There are some important risks associated with making such extensions
+incorrectly, so this article will try to present some guidance.
+
+### The proper namespace
+
+Interface points that are functions are usually designed to be found by
+argument-dependent lookup (ADL, see [TotW 49](/tips/49)). Operators and some
+operator-like functions (notably `swap`) are designed to be found by ADL. This
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: swap()

File: _posts/2017-11-16-totw-99.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+void bad();
+
+void test(const library::Letter& x) {
+  good(x);  // ok: 'library::good' is found by ADL.
+  bad(x);  // oops: '::bad' is hidden by 'client::bad'.
+}
+}  // namespace client
+```
+
+Note the difference between `library::good` and `::bad`. The `test` function is
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: paren all functions

File: _posts/2017-11-16-totw-99.md [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+relying on the absence of any function called `bad` in namespaces enclosing the
+call site. The appearance of `client::bad` hides `::bad` from the `test` caller.
+Meanwhile, the `good` function is found regardless of what else exists in the
+`test` function's enclosing scope. The C++ name lookup sequence will only yield
+a global if a name search from closer lexical scopes to the call site fails to
+find any names.
+
+This is all very subtle, which is the point, really. It's all much simpler if we
+default to defining functions alongside the data on which they operate.
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: Worth emphasizing that last line?

*default to defining functions alongside the data on which they operate*

File: _posts/2017-11-16-totw-99.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    swap(a.s_, b.s_);
+  }
+
+ private:
+  string s_;
+};
+}  // namespace library
+```
+
+They have a special property of ONLY being visible through ADL. They're a little
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: s/They/these friend functions

File: _posts/2017-11-16-totw-99.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+strange in that they are defined in the enclosing namespace, but don't appear in
+name lookup there. These in-class `friend` definitions must have inlined bodies
+for this stealth property to kick in. See [friend
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: Would capitalize, pluarlize, and enclose in quotes: See ["Friend Definitions"]

File: _posts/2017-11-16-totw-99.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+site is better. The important thing is that the specialization should not be
+hidden in client code: it should be as visible as the template and the
+user-defined types involved.
+
+### When to customize
+
+Don't customize the behavior of a class in a test. This is dangerous and
+unfortunately very common. Test source files are not immune to these dangers and
+should follow the same rules as production code. We find a lot of inappropriate
+operators in test.cc files that are written with the intent to "get `EXPECT_EQ`
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: \`*_test.cc\`

File: _posts/2017-11-16-totw-99.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+It's possible to define some customizations for raw pointers. It may be tempting
+in some cases to supply these customizations for `T*` along with `T`. This is
+not advised. It's dangerous because the customization may conflict with the
+expected ordinary behavior of pointers (e.g. the way they're logged, swapped, or
+compared). It's best to leave pointers alone.
+
+### What to do when you're stuck
+
+Following these guidelines can be challenging. Much of the inappropriate
+overloading and specialization seen in C++ code is motivated by small set of
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: by small -> by a small

File: _posts/2017-11-16-totw-99.md [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+root causes. Following is a partial listing with successful workarounds. If you
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: s/Following/Below

File: _posts/2017-11-16-totw-99.md [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+run into a library API that can't be worked with, please send a note to the
+owners to see about adding the appropriate customization hooks. Common APIs should
+be usable without breaking these interface packaging guidelines.
+
+#### Testing a type with `EXPECT_EQ`, etc.
+
+*temptation:* `EXPECT_EQ` requires `operator==`, and `operator<<` or gUnit's
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: I think the external spelling is "GUnit" but should we link to this?

(I can't find the aforementioned "PrintTo")

[GUnit](https://github.com/cpp-testing/GUnit)

File: _posts/2017-11-16-totw-99.md [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/99
+type: markdown
+order: "099"
+---
+
+Originally posted as totw/99 on 2015-06-24
+
+By a former Googler
```
Comment (manshreck) [inline _posts/2017-11-16-totw-99.md]: Omit (?) We should standardize, whatever the case.

File: _posts/2017-11-16-totw-99.md (hunk 3 above)
Comment (tituswinters) [reply _posts/2017-11-16-totw-99.md]: namespaces aren't indented.

File: _posts/2017-11-16-totw-99.md (hunk 12 above)
Comment (tituswinters) [reply _posts/2017-11-16-totw-99.md]: I'm not gonna worry about linking everything. PrintTo exists publicly, that's good enough for me for the moment.

**Comment by tituswinters** [comment]
Updated

**Pull request merged**